Kelompok kami menggunakan data dari Kaggle yang berjudul Data Train yang di mana kami akan mencoba untuk memprediksi harga tiket pesawat menggunakan metode Regresi Linear

###  1. Importing

Pengimportan library yang diperlukan

In [1]:
import pandas as pd
import numpy as np
import pandas as pd 
from matplotlib import pyplot as plt
import seaborn as sns
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error 
%matplotlib inline

 Di bawah ini adalah tampilan data dari Data Train yang kami dapat dari Kaggle. 

In [2]:
df = pd.read_csv('Data_train.csv', sep=";")
df['Date_of_Journey'] = pd.to_datetime(df['Date_of_Journey'])
display(df)

C:\Users\prisc\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '24/03/2019' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cache_array = _maybe_cache(arg, format, cache, convert_listlike)
C:\Users\prisc\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '24/06/2019' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cache_array = _maybe_cache(arg, format, cache, convert_listlike)
C:\Users\prisc\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '27/05/2019' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cache_array = _maybe_cache(arg, format, cache, convert_listlike)
C:\Users\prisc\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '18/04/2019' in DD/MM/YYYY format. Provide format or specif

Airline Date_of_Journey    Source Destination  \
0           IndiGo      2019-03-24  Banglore   New Delhi   
1        Air India      2019-01-05   Kolkata    Banglore   
2      Jet Airways      2019-09-06     Delhi      Cochin   
3           IndiGo      2019-12-05   Kolkata    Banglore   
4           IndiGo      2019-01-03  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia      2019-09-04   Kolkata    Banglore   
10679    Air India      2019-04-27   Kolkata    Banglore   
10680  Jet Airways      2019-04-27  Banglore       Delhi   
10681      Vistara      2019-01-03  Banglore   New Delhi   
10682    Air India      2019-09-05     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

### 2.1 Data Cleaning

Pembersihan data yang duplikat ataupun null pada data yang diberikan

In [3]:
#clean duplicated and null data
df.dropna(inplace=True)
df.drop_duplicates(keep='first',inplace=True)
df.shape

(10462, 11)

### 2.2 Date and Time

pemisahan data antara hari dan bulan

In [4]:
df['journey_day'] = df['Date_of_Journey'].dt.day
df['journey_month'] = df['Date_of_Journey'].dt.month
df['date_is_weekend'] = np.where(df['Date_of_Journey'].dt.day_name().isin(['Saturday','Sunday']),1,0)
df

Airline Date_of_Journey    Source Destination  \
0           IndiGo      2019-03-24  Banglore   New Delhi   
1        Air India      2019-01-05   Kolkata    Banglore   
2      Jet Airways      2019-09-06     Delhi      Cochin   
3           IndiGo      2019-12-05   Kolkata    Banglore   
4           IndiGo      2019-01-03  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia      2019-09-04   Kolkata    Banglore   
10679    Air India      2019-04-27   Kolkata    Banglore   
10680  Jet Airways      2019-04-27  Banglore       Delhi   
10681      Vistara      2019-01-03  Banglore   New Delhi   
10682    Air India      2019-09-05     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  journey_day  journey_month  date_is_weekend  
0             No info   3897           24              3                1  
1             No info   7662            5              1                1  
2             No info  13882            6              9                0  
3             No info   6218            5             12                0  
4             No info  13302            3              1                0  
...               ...    ...          ...            ...              ...  
10678         No info   4107            4              9                0  
10679         No info   4145           27              4                1  
10680         No info   7229           27              4                1  
10681         No info  12648            3              1                0  
10682         No info  11753            5              9                0  

[10462 rows x 14 columns]

### 2.3 Deploy and Arrival Hours

pemisahan antara jam dan menit pada keberangkatan dan kedatangan

In [5]:
#cleaning the deploy hour
df['Dep_Hour'] = df['Dep_Time'].apply(lambda x: int(x.split(':')[0]))
df['Dep_Min'] = df['Dep_Time'].apply(lambda x: int(x.split(':')[1]))

#cleaning the arrival hour
df['Arrival_Hour'] = df['Arrival_Time'].apply(lambda x: int(x.split(':')[0]))
df['Arrival_Min'] = df['Arrival_Time'].apply(lambda x: int(x.split(':')[1].split()[0]))

df

Airline Date_of_Journey    Source Destination  \
0           IndiGo      2019-03-24  Banglore   New Delhi   
1        Air India      2019-01-05   Kolkata    Banglore   
2      Jet Airways      2019-09-06     Delhi      Cochin   
3           IndiGo      2019-12-05   Kolkata    Banglore   
4           IndiGo      2019-01-03  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia      2019-09-04   Kolkata    Banglore   
10679    Air India      2019-04-27   Kolkata    Banglore   
10680  Jet Airways      2019-04-27  Banglore       Delhi   
10681      Vistara      2019-01-03  Banglore   New Delhi   
10682    Air India      2019-09-05     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  journey_day  journey_month  date_is_weekend  \
0             No info   3897           24              3                1   
1             No info   7662            5              1                1   
2             No info  13882            6              9                0   
3             No info   6218            5             12                0   
4             No info  13302            3              1                0   
...               ...    ...          ...            ...              ...   
10678         No info   4107            4              9                0   
10679         No info   4145           27              4                1   
10680         No info   7229           27              4                1   
10681         No info  12648            3              1                0   
10682         No info  11753            5              9                0   

       Dep_Hour  Dep_Min  Arrival_Hour  Arrival_Min  
0            22       20             1           10  
1             5       50            13           15  
2             9       25             4           25  
3            18        5            23           30  
4            16       50            21           35  
...         ...      ...           ...          ...  
10678        19       55            22           25  
10679        20       45            23           20  
10680         8       20            11           20  
10681        11       30            14           10  
10682        10       55            19           15  

[10462 rows x 18 columns]

### 2.4 Durations

pemisahan antara jam dan menit pada durasi

In [6]:
df["Duration"].unique()

def impute_Duration(duration):
    x = duration.split()
    if len(x) != 2:
        if 'h' in x[0]:
            return x[0] + ' 0m'
        else:
            return '0h ' + x[0]
    else:
        return duration
    
df['Duration'] = df['Duration'].apply(impute_Duration)

df['Duration_Hour'] = df['Duration'].apply(lambda x:int(x.split('h')[0]))
df['Duration_Minute'] = df['Duration'].apply(lambda x:x.split('m')[0].split()[1])
df

Airline Date_of_Journey    Source Destination  \
0           IndiGo      2019-03-24  Banglore   New Delhi   
1        Air India      2019-01-05   Kolkata    Banglore   
2      Jet Airways      2019-09-06     Delhi      Cochin   
3           IndiGo      2019-12-05   Kolkata    Banglore   
4           IndiGo      2019-01-03  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia      2019-09-04   Kolkata    Banglore   
10679    Air India      2019-04-27   Kolkata    Banglore   
10680  Jet Airways      2019-04-27  Banglore       Delhi   
10681      Vistara      2019-01-03  Banglore   New Delhi   
10682    Air India      2019-09-05     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun   19h 0m     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20    3h 0m    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  journey_day  journey_month  date_is_weekend  \
0             No info   3897           24              3                1   
1             No info   7662            5              1                1   
2             No info  13882            6              9                0   
3             No info   6218            5             12                0   
4             No info  13302            3              1                0   
...               ...    ...          ...            ...              ...   
10678         No info   4107            4              9                0   
10679         No info   4145           27              4                1   
10680         No info   7229           27              4                1   
10681         No info  12648            3              1                0   
10682         No info  11753            5              9                0   

       Dep_Hour  Dep_Min  Arrival_Hour  Arrival_Min  Duration_Hour  \
0            22       20             1           10              2   
1             5       50            13           15              7   
2             9       25             4           25             19   
3            18        5            23           30              5   
4            16       50            21           35              4   
...         ...      ...           ...          ...            ...   
10678        19       55            22           25              2   
10679        20       45            23           20              2   
10680         8       20            11           20              3   
10681        11       30            14           10              2   
10682        10       55            19           15              8   

      Duration_Minute  
0                  50  
1                  25  
2                   0  
3                  25  
4                  45  
...               ...  
10678              30  
10679              35  
10680               0  
10681              40  
10682              20  

[10462 rows x 20 columns]

### 2.5 Total Stops

pengintegrasian dari banyaknya transit

In [7]:
df['Total_Stops'].unique()
df = df.replace({'non-stop':0,'1 stop':1,'2 stops':2,'3 stops':3,'4 stops':4})

df = df.drop(['Route','Date_of_Journey','Dep_Time','Arrival_Time','Duration','Additional_Info'], axis='columns')
df

Airline    Source Destination  Total_Stops  Price  journey_day  \
0           IndiGo  Banglore   New Delhi            0   3897           24   
1        Air India   Kolkata    Banglore            2   7662            5   
2      Jet Airways     Delhi      Cochin            2  13882            6   
3           IndiGo   Kolkata    Banglore            1   6218            5   
4           IndiGo  Banglore   New Delhi            1  13302            3   
...            ...       ...         ...          ...    ...          ...   
10678     Air Asia   Kolkata    Banglore            0   4107            4   
10679    Air India   Kolkata    Banglore            0   4145           27   
10680  Jet Airways  Banglore       Delhi            0   7229           27   
10681      Vistara  Banglore   New Delhi            0  12648            3   
10682    Air India     Delhi      Cochin            2  11753            5   

       journey_month  date_is_weekend  Dep_Hour  Dep_Min  Arrival_Hour  \
0                  3                1        22       20             1   
1                  1                1         5       50            13   
2                  9                0         9       25             4   
3                 12                0        18        5            23   
4                  1                0        16       50            21   
...              ...              ...       ...      ...           ...   
10678              9                0        19       55            22   
10679              4                1        20       45            23   
10680              4                1         8       20            11   
10681              1                0        11       30            14   
10682              9                0        10       55            19   

       Arrival_Min  Duration_Hour Duration_Minute  
0               10              2              50  
1               15              7              25  
2               25             19               0  
3               30              5              25  
4               35              4              45  
...            ...            ...             ...  
10678           25              2              30  
10679           20              2              35  
10680           20              3               0  
10681           10              2              40  
10682           15              8              20  

[10462 rows x 14 columns]

### 2.6 Removing Outliers

In [8]:
def remove_airline_price_outliers(df):
    df_out = pd.DataFrame()
    for airline,airline_df in df.groupby('Airline'):
        for source,source_df in airline_df.groupby('Source'):
            for destination,destination_df in source_df.groupby('Destination'):
                mean = np.mean(destination_df['Price'])
                stdev = np.std(destination_df['Price'])
                reduced_df = destination_df[(destination_df['Price'] >=(mean-stdev)) & (destination_df['Price'] <= (mean+stdev))]
                df_out = pd.concat([df_out,reduced_df],ignore_index=True)
    return df_out

df = remove_airline_price_outliers(df)

### 3.1 Model Building

In [9]:
x = df.drop('Price',axis=1)
y = df['Price']
dummies = pd.get_dummies(x[['Airline','Source','Destination']],drop_first=True)
x = pd.concat([x, dummies], axis=1)
x = x.drop(["Airline", "Source", "Destination"], axis= "columns")
display(x)

x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=42)

model = DecisionTreeRegressor(max_depth=30,min_samples_leaf=1,min_samples_split=18,splitter='random')

#model training
model.fit(x_train, y_train)
print(f"Model Accuration {model.score(x_test,y_test)}")

Total_Stops  journey_day  journey_month  date_is_weekend  Dep_Hour  \
0               0            5              6                0        11   
1               0            4              3                0        23   
2               0            6             12                0        11   
3               0           24              4                0        11   
4               0           18              4                0        11   
...           ...          ...            ...              ...       ...   
7879            1            3              6                0        20   
7880            1            3              6                0        10   
7881            0            4              6                0        11   
7882            0            3              3                1        16   
7883            0            3              1                0         7   

      Dep_Min  Arrival_Hour  Arrival_Min  Duration_Hour Duration_Minute  ...  \
0          10            13           55              2              45  ...   
1          55             2           45              2              50  ...   
2          10            13           55              2              45  ...   
3          10            13           55              2              45  ...   
4          10            13           55              2              45  ...   
...       ...           ...          ...            ...             ...  ...   
7879       55            22           55             26               0  ...   
7880       25            22           55             12              30  ...   
7881       30            14           20              2              50  ...   
7882        0            18           35              2              35  ...   
7883        5             9           20              2              15  ...   

      Airline_Vistara Premium economy  Source_Chennai  Source_Delhi  \
0                                   0               0             0   
1                                   0               0             0   
2                                   0               0             0   
3                                   0               0             0   
4                                   0               0             0   
...                               ...             ...           ...   
7879                                0               0             0   
7880                                0               0             0   
7881                                1               0             0   
7882                                1               0             0   
7883                                1               1             0   

      Source_Kolkata  Source_Mumbai  Destination_Cochin  Destination_Delhi  \
0                  0              0                   0                  1   
1                  0              0                   0                  1   
2                  0              0                   0                  1   
3                  0              0                   0                  1   
4                  0              0                   0                  1   
...              ...            ...                 ...                ...   
7879               0              1                   0                  0   
7880               0              1                   0                  0   
7881               0              0                   0                  1   
7882               0              0                   0                  0   
7883               0              0                   0                  0   

      Destination_Hyderabad  Destination_Kolkata  Destination_New Delhi  
0                         0                    0                      0  
1                         0                    0                      0  
2                         0                    0                      0  
3                         0                   

Model Accuration 0.8897326658164064


Di atas adalah tampilan data yang sudah dibersihkan dan sudah melalui proses One Hot Encoding yaitu teknik di mana suatu fitur bertipe data kategorial dengan  𝑘 buah kategori berbeda akan 'dimekarkan' menjadi  𝑘 buah fitur baru.

### 3.2 Model Testing

In [10]:
testing_data = x[-1:]
[prediction_data] = model.predict(testing_data)
print(f"Hasil prediksi dari model kami {prediction_data}")
print(f"harga dari da9125taset kami {y[-1:].to_string(index=False, header=False)}")

Hasil prediksi dari model kami 9125.0
harga dari da9125taset kami 9125


In [11]:
#MSE data latih
y_train_pred = model.predict(x_train)
mse_train = mean_squared_error(y_train, y_train_pred)
print(f"MSE pada data latih: {mse_train}")

#MSE data uji
y_test_pred = model.predict(x_test)
mse_test = mean_squared_error(y_test, y_test_pred)
print(f"MSE pada data uji: {mse_test}")

MSE pada data latih: 1141370.2445972017
MSE pada data uji: 1522562.243609724


In [12]:
#MSA data latih
y_train_pred = model.predict(x_train)
msa_train = mean_absolute_error(y_train, y_train_pred)
print(f"MSA pada data latih: {msa_train}")

#MSA data uji
y_test_pred = model.predict(x_test)
msa_test = mean_absolute_error(y_test, y_test_pred)
print(f"MSA pada data uji: {msa_test}")

MSA pada data latih: 680.5977290680067
MSA pada data uji: 795.3207092968476


# Kesimpulan 

Berdasarkan langkah-langkah yang telah kami lakukan dari pembersihan data, pemisahan beberapa kategori, menghapus data yang menjadi pencilan, membangun model regresi, melakukan prediksi dan di akhir memeriksa fungsi biaya latih. Dari perhitungan dapat dilihat bahwa fungsi biaya pada data uji yaitu 795.3207092968476 sudah lebih baik daripada fungsi biaya pada data latih yaitu 680.5977290680067. Artinya sistem prediksi sudah dapat berjalan dengan baik.